### ISSUE TITLE
Panic on tuple struct

### ISSUE THREAD

**Issue by Yamakaky** [serde-rs/serde#572]
failed to parse macro input: "pub struct Screen(pub [Color; 12288 as usize]);"

It works without the `as usize`.

**Comment by dtolnay** [comment]
Thanks for the report! Is there a specific reason you wanted to use "as usize"? Would it be reasonable for you to use one of these alternatives, both of which are supported?

```rust
pub struct Screen(pub [Color; 12288]);
pub struct Screen(pub [Color; 12288usize]);
```

This was a conscious decision not to support arbitrary expressions as array lengths. Rust's type grammar is pretty wild. For example [you can define functions inside of your type](https://is.gd/52Ygtk):

```rust
pub struct Screen(pub [Color; {
    fn holy_smokes() {
        println!("why am I part of this type?");
    }
    12288
}]);
```

The advantage of supporting integers only is that `syn` compiles more than twice as fast compared to supporting expressions. That said, if there is a subset of expressions that are commonly used as array lengths for a compelling reason then we will need to add support for that subset.

As a separate issue I filed https://github.com/dtolnay/syn/issues/28 upstream to improve the error message when an array length expression is not supported.

**Comment by Yamakaky** [comment]
In fact, I was using `SCREEN_SIZE as usize`.

**Comment by dtolnay** [comment]
That does seem like a compelling use case. I filed https://github.com/dtolnay/syn/issues/29 to track this.

**Comment by dtolnay** [comment]
The parsing issue is fixed in 0.8.12.

**Comment by Yamakaky** [comment]
cool!
